### PR TITLE
feat: Add ODP Segments support in Audience Evaluation

### PR DESCRIPTION
**Pull request by mnoman09**
## Summary
1. Added ODP Segments support in Audience evaluation
2. Adding logic to parse Integrations from datafile and make odp host and key available in projectConfig.

## Test plan
- Manually tested thoroughly.
- Added New unit tests.
- Existing unit tests pass.
- Existing Full stack compatibility tests pass.

## JIRA
[OASIS-8382](https://optimizely.atlassian.net/browse/OASIS-8382)

### CODE REVIEW

File: core-api/src/main/java/com/optimizely/ab/config/audience/UserAttribute.java [hunk 1]
```diff
@@ -71,19 +75,29 @@ public Object getValue() {
...
 
-        if (!"custom_attribute".equals(type)) {
+        if (!isValidType(type)) {
             logger.warn("Audience condition \"{}\" uses an unknown condition type. You may need to upgrade to a newer release of the Optimizely SDK.", this);
             return null; // unknown type
         }
         // check user attribute value is equal
         try {
+            // Handle qualified segments
+            if ("qualified".equals(match)) {
```
Comment (zashraf1985) [inline core-api/src/main/java/com/optimizely/ab/config/audience/UserAttribute.java:93]: This string literal should be moved to a constant

File: core-api/src/main/java/com/optimizely/ab/config/audience/UserAttribute.java [hunk 2]
```diff
@@ -118,6 +132,13 @@ public Boolean evaluate(ProjectConfig config, Map<String, ?> attributes) {
         return null;
     }
 
+    private boolean isValidType(String type) {
+        if (type.equals(CUSTOM_ATTRIBUTE.toString()) || type.equals(THIRD_PARTY_DIMENSION.toString())) {
```
Comment (zashraf1985) [inline core-api/src/main/java/com/optimizely/ab/config/audience/UserAttribute.java:136]: this is not scalable. adding more types will result in more explicit conditions in this check. may be add the types to a list and check for existence?

**Review by zashraf1985 — DISMISSED** [review]

File: core-api/src/main/java/com/optimizely/ab/config/DatafileProjectConfig.java [hunk 3]
```diff
@@ -63,6 +63,8 @@ public class DatafileProjectConfig implements ProjectConfig {
     private final boolean anonymizeIP;
     private final boolean sendFlagDecisions;
     private final Boolean botFiltering;
+    private final String hostForODP;
+    private final String publicKeyForODP;
```
Comment (mikechu-optimizely) [inline core-api/src/main/java/com/optimizely/ab/config/DatafileProjectConfig.java:67]: For acronyms in names, are we using `ODP` or `Odp` ?

File: .github/workflows/java.yml [hunk 4]
```diff
@@ -45,7 +45,7 @@ jobs:
   
   test:
     if: startsWith(github.ref, 'refs/tags/') != true
-    runs-on: ubuntu-18.04
+    runs-on: macos-latest
```
Comment (jaeopt) [inline .github/workflows/java.yml:48]: why macos?

File: core-api/src/main/java/com/optimizely/ab/config/Integration.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        this.key = key;
+        this.host = host;
+        this.publicKey = publicKey;
+    }
+
+    public String getKey() {
+        return key;
+    }
+
+    public String getHost() { return host; }
```
Comment (jaeopt) [inline core-api/src/main/java/com/optimizely/ab/config/Integration.java:50]: @Nullable

File: core-api/src/main/java/com/optimizely/ab/config/Integration.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    public String getPublicKey() { return publicKey; }
```
Comment (jaeopt) [inline core-api/src/main/java/com/optimizely/ab/config/Integration.java:52]: @Nullable

File: core-api/src/main/java/com/optimizely/ab/config/Integration.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    @JsonCreator
+    public Integration(@JsonProperty("key") String key,
+                       @JsonProperty("host") String host,
+                       @JsonProperty("publicKey") String publicKey) {
+        this.key = key;
+        this.host = host;
+        this.publicKey = publicKey;
+    }
+
+    public String getKey() {
```
Comment (jaeopt) [inline core-api/src/main/java/com/optimizely/ab/config/Integration.java:46]: @Nonnull

File: core-api/src/main/java/com/optimizely/ab/config/Integration.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    public String getHost() { return host; }
+
+    public String getPublicKey() { return publicKey; }
+
+    @Override
+    public String toString() {
+        return "Integration{" +
+            "key='" + key + '\'' +
+            ", host='" + host + '\'' +
+            ", publicKey='" + publicKey + '\'' +
```
Comment (jaeopt) [inline core-api/src/main/java/com/optimizely/ab/config/Integration.java:59]: Can we make this toString for generic integration types, not only for ODP?
Like "if (host != null) { string += host }..."

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks good! A few small changes suggested.
Also missing tests for validating integrations (no integrations, empty integrations, key-only integration, key-missing integration, etc). See https://github.com/optimizely/swift-sdk/blob/jae/ats/Tests/OptimizelyTests-DataModel/IntegrationTests.swift

File: core-api/src/main/java/com/optimizely/ab/config/Integration.java (hunk 6 above)
Comment (zashraf1985) [reply core-api/src/main/java/com/optimizely/ab/config/Integration.java:52]: I think it is nullable in java unless specified the other way round?

File: .github/workflows/java.yml (hunk 4 above)
Comment (zashraf1985) [reply .github/workflows/java.yml:48]: Because some tests were failing for no apparent reason on linux. we have changes it to macox for now. will fix it later as part of a separate ticket

**Review by jaeopt — CHANGES_REQUESTED** [review]
All changes look good except that I do not see integration tests. Missing?

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks good. One small change suggested.

File: core-api/src/test/java/com/optimizely/ab/config/parser/GsonConfigParserTest.java [hunk 9]
```diff
@@ -313,6 +313,52 @@ public void nullJsonExceptionWrapping() throws Exception {
...
+        assertEquals(actual.getHostForODP(), "https://example.com");
+        assertEquals(actual.getPublicKeyForODP(), "test-key");
+    }
+
+    @Test
+    public void integrationsArrayHasOtherIntegration() throws Exception {
+        GsonConfigParser parser = new GsonConfigParser();
+        String integrationsObject = ", \"integrations\": [" +
+            "{ \"key\": \"not-odp\", " +
+            "\"host\": \"https://example.com\", " +
```
[the remark this answers is not in the document]
Comment (zashraf1985) [reply core-api/src/test/java/com/optimizely/ab/config/parser/GsonConfigParserTest.java:337]: Done

**Review by jaeopt — APPROVED** [review]
LGTM
